### PR TITLE
Big refactor

### PR DESCRIPTION
**Pull request by Anya497**
- Refactor statistics collector. Replace ASCII table with `csv` file for maps results
- Log single SVM average coverage and number of failed maps to mlflow
- Refactor `run_training.py`. Move validation function initialization to `run_training` function. `objective`  function responsible only for running `training-validation` cycle regardless of validation type
- Move `fail_immediately` flag and `PlatformsConfig` to `ValidationWithSVMs` because they are necessary only for this validation type

### CODE REVIEW

File: AIAgent/ml/training/validation.py [hunk 1]
```diff
@@ -82,33 +82,17 @@ def validate_coverage(
             if isinstance(result, GameError):
                 need_to_save_map: bool = result.need_to_save_map()
                 if not need_to_save_map:
-                    statistics_collector.fail(result._map)
-            else:
-                all_results.append(result)
-    statistics_collector.update_results(all_results)
+                    maps_to_remove.append(result.map2result.map)
```
Comment (gsvgit) [inline AIAgent/ml/training/validation.py:85]: `maps_to_remove` will be used to check whether should we fail immediately but here we use additional flag `need_to_save_map` to fill this collection. Looks strange.

File: AIAgent/run_training.py [hunk 2]
```diff
@@ -70,72 +69,104 @@ class TrialSettings:
...
-                for _map in single_json_maps:
-                    fullName = os.path.join(dataset_base_path, _map.AssemblyFullName)
-                    _map.AssemblyFullName = fullName
-                    single_json_maps_with_svms.append(GameMap2SVM(_map, svm_info))
-                maps.extend(single_json_maps_with_svms)
+        def validate(model, dataset: TrainingDataset):
+            result, metrics, maps_to_remove = validate_coverage(
+                model, dataset, maps, validation_config.validation
+            )
+            if maps_to_remove and validation_config.validation.fail_immediately:
```
Comment (gsvgit) [inline AIAgent/run_training.py:127]: Look at [this code](https://github.com/PySymGym/PySymGym/pull/120/files#diff-3f4e39c863773963ee6442034b7b73bcac4f3003ea83f881cdafaf824ecf3be7R85). `fail_immediate`  must be independent from other flags. CI must fails on any problems with maps in any cases.

File: AIAgent/run_training.py [hunk 3]
```diff
@@ -70,72 +69,104 @@ class TrialSettings:
...
+                dataset,
+                criterion_init(),
+                validation_config.validation.batch_size,
+            )
+    elif isinstance(validation_config.validation, ValidationWithSVMs):
+
+        def load_maps_info():
+            maps: list[GameMap2SVM] = list()
+            for platform in validation_config.validation.platforms_config:
+                svms_info = platform.svms_info
```
Comment (gsvgit) [inline AIAgent/run_training.py:94]: Can be inlined.

File: AIAgent/run_training.py [hunk 4]
```diff
@@ -70,72 +69,104 @@ class TrialSettings:
...
+                for svm_info in svms_info:
+                    for dataset_config in platform.dataset_configs:
+                        dataset_base_path = dataset_config.dataset_base_path
```
Comment (gsvgit) [inline AIAgent/run_training.py:97]: The body can be moved to top-level function, after that residual code can be inlined as list generator.

**Review by gsvgit — CHANGES_REQUESTED** [review]

**Review by Parzival-05 — CHANGES_REQUESTED** [review]

File: AIAgent/ml/training/validation.py [hunk 5]
```diff
@@ -80,35 +80,21 @@ def validate_coverage(
             colour=progress_bar_colour,
         ):
             if isinstance(result, GameError):
+                if validation_config.fail_immediately:
+                    raise RuntimeError("Validation failed")
```
[the remark this answers is not in the document]
Comment (gsvgit) [reply AIAgent/ml/training/validation.py:84]: I think, we can return to this question back later an fix it separately.

File: AIAgent/ml/training/utils.py [hunk 6]
```diff
@@ -149,3 +151,23 @@ def remove_extra_attrs(step: HeteroData):
...
+    for map2result in map2result_list:
+        _map = map2result.map
+        map_name = _map.GameMap.MapName
+        game_result_str = map2result.game_result.printable(verbose=True)
+        maps.append(f"{_map.SVMInfo.name} : {map_name}")
+        results.append(game_result_str)
+
+    df = pd.DataFrame(results, columns=["Game result"], index=maps).T
+
+    return df
```
[the remark this answers is not in the document]
Comment (gsvgit) [reply AIAgent/ml/training/utils.py:173]: Well... But utils reorganization is a separated task.

File: AIAgent/ml/training/validation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Anya497) [reply AIAgent/ml/training/validation.py:84]: I've created the [issue](https://github.com/PySymGym/PySymGym/issues/130).

File: AIAgent/ml/training/statistics.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            )
+            metrics[f"average_coverage_for_{svm_info.name}"] = avg_by_attr(
+                [
+                    map2result.game_result
+                    for map2result in successful_maps
+                    if map2result.map.SVMInfo.name == svm_info.name
+                ],
+                "actual_coverage_percent",
+            )
+    return metrics["average_coverage"], metrics
```
[the remark this answers is not in the document]
Comment (Anya497) [reply AIAgent/ml/training/statistics.py:61]: We need separate metric which is used for tuning hyperparameters with `optuna`. I think it's better define it here.

File: AIAgent/ml/training/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Anya497) [reply AIAgent/ml/training/utils.py:173]: These functions can be useful for some special experiments or for analysis of the results. They are not the part of training pipeline and this is the reason of their location in `utils`.

File: AIAgent/ml/training/statistics.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        for map2result in results
+        if not isinstance(map2result.game_result, GameFailed)
+    ]
+    metrics = {
+        "average_dataset_state_coverage": avg_by_attr(
+            dataset.maps_results.values(), "coverage_percent"
+        ),
+        "average_coverage": avg_by_attr(
+            [map2result.game_result for map2result in successful_maps],
+            "actual_coverage_percent",
```
Comment (gsvgit) [inline AIAgent/ml/training/statistics.py:54]: `"actual_coverage_percent"` used at least twice. Move it to (local) constant.

File: AIAgent/ml/training/statistics.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gsvgit) [reply AIAgent/ml/training/statistics.py:74]: Look at https://github.com/PySymGym/PySymGym/pull/120/files#diff-c197b7ad994b864760fb400630cf61a275cb411f3ae6eacd3fe8de399fa54e1eR131

File: AIAgent/tests/test_statistics.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        self.tmp_dir.mkdir(exist_ok=True)
+        with open(self.test_csv_file_path, "w") as statistics_file:
+            statistics_writer = csv.DictWriter(
+                statistics_file,
+                sorted([game_map2svm.GameMap.MapName for game_map2svm in maps]),
+            )
+            statistics_writer.writeheader()
+
+    def remove_tmp(self):
+        shutil.rmtree("tests/tmp")
```
Comment (gsvgit) [inline AIAgent/tests/test_statistics.py:34]: Move to constant

File: AIAgent/tests/test_statistics.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                "Map1": Result(100, -4, -9028, 5),
+                "Map2": Result(0, -1, -8, 0),
+            }
+        )
+        maps = get_maps(val_config)
+        return val_config, dataset, maps
+
+    def test_single_epoch_with_successful_results(self, get_args):
+        metrics = {
+            "average_coverage": 100.0,
```
Comment (gsvgit) [inline AIAgent/tests/test_statistics.py:60]: Can we generate names using svms?
Can we move svm-independent names to constants to reuse  [here](https://github.com/PySymGym/PySymGym/pull/120/files#diff-77b01a395331db59634890091292862d8a6cdbaaf4de890c56bbc4c67d845667R49)

File: AIAgent/common/classes.py [hunk 11]
```diff
@@ -45,7 +63,7 @@ class Agent2Result:
 @dataclass
 class Map2Result:
     map: GameMap2SVM
-    game_result: GameResult
+    game_result: Union[GameResult, GameFailed]
```
Comment (Parzival-05) [inline AIAgent/common/classes.py:66]: 👍

File: AIAgent/ml/game/play_game.py [hunk 12]
```diff
@@ -164,30 +162,35 @@ def play_game(
...
             )
-        logging.warning(log_message)
+        )
+        game_result = GameFailed(reason=type(error))
 
-        if need_to_save:
-            FeatureConfig.SAVE_IF_FAIL_OR_TIMEOUT.save_model(
-                with_predictor.model(), with_name=name_of_predictor
-            )
-        raise GameError(game_map2svm=game_map2svm, error_name=error.__class__.__name__)
```
Comment (Parzival-05) [inline AIAgent/ml/game/play_game.py:192]: Copypasta

File: AIAgent/tests/resources/svms_validation_configs/real_svms_config.yml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    - name: java
+      DatasetConfigs:
+        - dataset_base_path: ../maps/Java/Maps/
+          dataset_description: ../maps/Java/Maps/dataset.json
+      SVMSInfo:
+        - name: usvm
+          launch_command: java -jar app.jar --mode server --port {port}
+          min_port: 35500
+          max_port: 35950
+          server_working_dir: ../GameServers/usvm/usvm-ml-gameserver/build/libs/
```
Comment (Parzival-05) [inline AIAgent/tests/resources/svms_validation_configs/real_svms_config.yml:24]: \n

File: AIAgent/ml/game/play_game.py (hunk 12 above)
Comment (Anya497) [reply AIAgent/ml/game/play_game.py:192]: I didn't understand you

File: AIAgent/ml/training/statistics.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                header,
+            )
+        )
+    )
+    with open(paths.CURRENT_TABLE_PATH, "a") as statistics_file:
+        statistics_writer = csv.DictWriter(
+            statistics_file, sorted(results_to_write.keys())
+        )
+        statistics_writer.writerow(results_to_write)
+    GAME_RESULT_COVERAGE_PERCENT_ATTR = "actual_coverage_percent"
```
Comment (gsvgit) [inline AIAgent/ml/training/statistics.py:42]: Can it be moved up, to collect all literals in one single place?

File: AIAgent/ml/training/statistics.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        if isinstance(map2result.game_result, GameFailed)
+    ]
+    successful_maps = [
+        map2result
+        for map2result in results
+        if not isinstance(map2result.game_result, GameFailed)
+    ]
+    metrics = {
+        AVERAGE_COVERAGE_OF_DATASET_STATE: avg_by_attr(
+            dataset.maps_results.values(), "coverage_percent"
```
Comment (gsvgit) [inline AIAgent/ml/training/statistics.py:55]: Can `"coverage_percent"` be declared as constant literal? Just for unification.

**Review by gsvgit — APPROVED** [review]

**Review by Parzival-05 — APPROVED** [review]

**Pull request merged**